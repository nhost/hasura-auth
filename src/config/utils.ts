// * Helpers for casting environment variables

export const castStringEnv = (envVar: string, defaultValue = ''): string => {
  return process.env[envVar] ? (process.env[envVar] as string) : defaultValue;
};

export const castBooleanEnv = (
  envVar: string,
  defaultValue = false
): boolean => {
  return process.env[envVar]
    ? process.env[envVar]?.toLowerCase() === 'true'
    : defaultValue;
};

export const castIntEnv = (envVar: string, defaultValue: number): number => {
  return process.env[envVar]
    ? parseInt(process.env[envVar] as string, 10)
    : defaultValue;
};

export const castStringArrayEnv = (
  envVar: string,
  defaultValue: string[] = []
): string[] => {
  return process.env[envVar]?.length
    ? (process.env[envVar] as string).split(',').map((field) => field.trim())
    : defaultValue;
};

export const castObjectEnv = <T extends Record<string, unknown>>(
  envVar: string,
  defaultValue: T = {} as T
): T => {
  const env = process.env[envVar];
  if (env) {
    return JSON.parse(env);
  } else return defaultValue;
};
