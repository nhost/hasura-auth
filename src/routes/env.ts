import { Router } from 'express';
import { sendError } from '@/errors';

export default (router: Router) => {
  // THESE ENDPOINTS ARE ONLY TO BE USED FOR TESTS!!
  // They allows us to programmatically enable/disable
  // functionality needed for specific tests.

  const envStack: unknown[] = [];

  router.post('/change-env', (req, res) => {
    if (process.env.NODE_ENV === 'production') {
      return sendError(res, 'forbidden-endpoint-in-production');
    }

    envStack.push(Object.assign({}, process.env));

    Object.assign(process.env, req.body);

    res.json(process.env);
  });

  router.get('/env/:id', (req, res) => {
    if (process.env.NODE_ENV === 'production') {
      return sendError(res, 'forbidden-endpoint-in-production');
    }

    res.send(process.env[req.params.id]);
  });
};
