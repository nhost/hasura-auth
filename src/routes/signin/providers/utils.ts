import express, {
  NextFunction,
  RequestHandler,
  Response,
  Request,
  Router,
} from 'express';
import passport, { Profile } from 'passport';
import { VerifyCallback } from 'passport-oauth2';
import refresh from 'passport-oauth2-refresh';
import { Strategy } from 'passport';
import { v4 as uuidv4 } from 'uuid';

import { PROVIDERS } from '@config';
import {
  Joi,
  email as emailValidator,
  uuid,
  queryValidator,
  registrationOptions,
} from '@/validation';
import { UserFieldsFragment } from '@/utils/__generated__/graphql-request';
import {
  asyncWrapper,
  getNewRefreshToken,
  gqlSdk,
  getUserByEmail,
  insertUser,
  getGravatarUrl,
  ENV,
} from '@/utils';
import { UserRegistrationOptions } from '@/types';

export const providerCallbackQuerySchema = Joi.object({
  state: uuid.required(),
}).unknown(true);

type ProviderQuery = UserRegistrationOptions & {
  redirectTo: string;
};

type ProviderCallbackQuery = Record<string, unknown> & {
  state: string;
};

type RequestWithState<Q = {}> = Request<{}, {}, {}, Q & { state: string }> & {
  state: string;
};

interface Constructable<T> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): T;
  prototype: T;
}

export type TransformProfileFunction = <T extends Profile>(
  profile: T
) => { id: string; email: string; displayName: string; avatarUrl: string };

interface InitProviderSettings {
  transformProfile: TransformProfileFunction;
  callbackMethod: 'GET' | 'POST';
}

const manageProviderStrategy =
  (provider: string, transformProfile: TransformProfileFunction) =>
  async (
    req: RequestWithState<ProviderCallbackQuery>,
    accessToken: string,
    refreshToken: string,
    profile: Profile,
    done: VerifyCallback
  ): Promise<void> => {
    const state = req.query.state;

    const requestOptions = await gqlSdk
      .providerRequest({
        id: state,
      })
      .then((res) => res.authProviderRequest?.options);

    // find or create the user
    // check if user exists, using profile.id
    const { id, email, displayName, avatarUrl } = transformProfile(profile);

    // check if user already exist with `id` (unique id from provider)
    const userProvider = await gqlSdk
      .authUserProviders({
        provider,
        providerUserId: id.toString(),
      })
      .then((res) => {
        return res.authUserProviders[0];
      });

    // User is already registered
    if (userProvider) {
      await gqlSdk.updateAuthUserprovider({
        id: userProvider.id,
        authUserProvider: {
          accessToken,
          refreshToken,
        },
      });

      return done(null, userProvider.user);
    }

    if (email) {
      try {
        await emailValidator.validateAsync(email);
      } catch {
        return done(new Error('email is not allowed'));
      }

      const user = await getUserByEmail(email);

      if (user) {
        // add this provider to existing user with the same email
        const insertedAuthUserprovider = await gqlSdk
          .insertUserProviderToUser({
            userProvider: {
              userId: user.id,
              providerId: provider,
              providerUserId: id.toString(),
              accessToken,
              refreshToken,
            },
          })
          .then((res) => res.insertAuthUserProvider);

        if (!insertedAuthUserprovider) {
          throw new Error('Could not insert provider to user');
        }

        return done(null, user);
      }
    }

    const { defaultRole, locale, allowedRoles, metadata } = requestOptions;

    const insertedUser = await insertUser({
      email,
      passwordHash: null,
      emailVerified: !!email,
      defaultRole,
      locale,
      roles: {
        data: (allowedRoles as string[]).map((role) => ({
          role,
        })),
      },
      displayName: requestOptions.displayName || displayName || email,
      avatarUrl,
      metadata,
      userProviders: {
        data: [
          {
            providerUserId: id.toString(),
            accessToken,
            refreshToken,
            providerId: provider,
          },
        ],
      },
    });

    return done(null, insertedUser);
  };

const providerCallback = asyncWrapper(
  async (req: RequestWithState, res: Response): Promise<void> => {
    // Successful authentication, redirect home.
    // generate tokens and redirect back home

    req.state = req.query.state as string;

    const requestOptions = await gqlSdk
      .deleteProviderRequest({
        id: req.state,
      })
      .then((res) => res.deleteAuthProviderRequest?.options);

    const user = req.user as UserFieldsFragment;

    const refreshToken = await getNewRefreshToken(user.id);

    // redirect back user to app url
    // ! temparily send the refresh token in both hash and query parameter
    // TODO at a later stage, only send as a query parameter
    res.redirect(
      `${requestOptions.redirectTo}?refreshToken=${refreshToken}#refreshToken=${refreshToken}`
    );
  }
);

export const initProvider = <T extends Strategy>(
  router: Router,
  strategyName:
    | 'github'
    | 'google'
    | 'facebook'
    | 'twitter'
    | 'linkedin'
    | 'apple'
    | 'windowslive'
    | 'spotify'
    | 'gitlab'
    | 'bitbucket'
    | 'strava'
    | 'discord'
    | 'twitch',
  strategy: Constructable<T>,
  settings: InitProviderSettings & ConstructorParameters<Constructable<T>>[0], // TODO: Strategy option type is not inferred correctly
  middleware?: RequestHandler
): void => {
  const {
    transformProfile = ({
      id,
      emails,
      displayName,
      photos,
    }: Profile): {
      id: string;
      email?: string;
      displayName: string;
      avatarUrl?: string;
    } => ({
      id,
      email: emails?.[0]?.value,
      displayName: displayName,
      avatarUrl: photos?.[0]?.value || getGravatarUrl(emails?.[0].value),
    }),
    callbackMethod = 'GET',
    ...options
  } = settings;

  const subRouter = Router();

  if (middleware) {
    subRouter.use(middleware);
  }

  if (PROVIDERS[strategyName]) {
    const strategyToUse = new strategy(
      {
        ...PROVIDERS[strategyName],
        ...options,
        callbackURL: `${ENV.AUTH_SERVER_URL}/signin/provider/${strategyName}/callback`,
        passReqToCallback: true,
      },
      manageProviderStrategy(strategyName, transformProfile)
    );

    passport.use(strategyName, strategyToUse);
    // @ts-expect-error
    refresh.use(strategyToUse);
  }

  subRouter.get('/', [
    queryValidator(registrationOptions),
    asyncWrapper(
      async (
        req: RequestWithState<ProviderQuery>,
        res: Response,
        next: NextFunction
      ) => {
        req.state = uuidv4();
        // insert request metadata object with the request state as id
        await gqlSdk.insertProviderRequest({
          providerRequest: {
            id: req.state,
            options: req.query,
          },
        });

        next();
      }
    ),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (req: RequestWithState<ProviderQuery>, ...rest: any) => {
      return passport.authenticate(strategyName, {
        session: false,
        state: req.state,
        ...options,
      })(req, ...rest);
    },
  ]);

  const handlers = [
    passport.authenticate(strategyName, {
      session: false,
    }),
    queryValidator(providerCallbackQuerySchema),
    providerCallback,
  ];

  if (callbackMethod === 'POST') {
    // The Sign in with Apple auth provider requires a POST route for authentication
    subRouter.post(
      '/callback',
      express.urlencoded({ extended: true }),
      ...handlers
    );
  } else {
    subRouter.get('/callback', ...handlers);
  }

  router.use(`/${strategyName}`, subRouter);
};
