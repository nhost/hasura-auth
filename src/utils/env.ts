import {
  castBooleanEnv,
  castIntEnv,
  castObjectEnv,
  castStringArrayEnv,
  castStringEnv,
} from '../config/utils';

export const ENV = {
  get PWD() {
    return castStringEnv('PWD', '/');
  },

  // HASURA
  get HASURA_GRAPHQL_JWT_SECRET() {
    return castStringEnv('HASURA_GRAPHQL_JWT_SECRET', '');
  },
  get HASURA_GRAPHQL_DATABASE_URL() {
    return castStringEnv('HASURA_GRAPHQL_DATABASE_URL', '');
  },
  get HASURA_GRAPHQL_ADMIN_SECRET() {
    return castStringEnv('HASURA_GRAPHQL_ADMIN_SECRET', '');
  },
  get HASURA_GRAPHQL_GRAPHQL_URL() {
    return castStringEnv('HASURA_GRAPHQL_GRAPHQL_URL', '');
  },

  // SERVER
  get AUTH_HOST() {
    return castStringEnv('AUTH_HOST', '0.0.0.0');
  },
  get AUTH_PORT() {
    return castIntEnv('AUTH_PORT', 4000);
  },
  get AUTH_SERVER_URL() {
    return castStringEnv('AUTH_SERVER_URL', '');
  },

  // SMTP
  get AUTH_SMTP_PASS() {
    return castStringEnv('AUTH_SMTP_PASS', '');
  },
  get AUTH_SMTP_HOST() {
    return castStringEnv('AUTH_SMTP_HOST', '');
  },
  get AUTH_SMTP_USER() {
    return castStringEnv('AUTH_SMTP_USER', '');
  },
  get AUTH_SMTP_SENDER() {
    return castStringEnv('AUTH_SMTP_SENDER', '');
  },
  get AUTH_SMTP_AUTH_METHOD() {
    return castStringEnv('AUTH_SMTP_AUTH_METHOD', 'PLAIN');
  },
  get AUTH_SMTP_PORT() {
    return castIntEnv('AUTH_SMTP_PORT', 587);
  },
  get AUTH_SMTP_SECURE() {
    // note: false disables SSL (deprecated)
    return castBooleanEnv('AUTH_SMTP_SECURE', false);
  },

  // SMS
  get AUTH_SMS_PROVIDER() {
    return castStringEnv('AUTH_SMS_PROVIDER', '');
  },

  // TWILIO
  get AUTH_SMS_TWILIO_ACCOUNT_SID() {
    return castStringEnv('AUTH_SMS_TWILIO_ACCOUNT_SID', '');
  },
  get AUTH_SMS_TWILIO_AUTH_TOKEN() {
    return castStringEnv('AUTH_SMS_TWILIO_AUTH_TOKEN', '');
  },
  get AUTH_SMS_TWILIO_MESSAGING_SERVICE_ID() {
    return castStringEnv('AUTH_SMS_TWILIO_MESSAGING_SERVICE_ID', '');
  },
  get AUTH_SMS_TWILIO_FROM() {
    return castStringEnv('AUTH_SMS_TWILIO_FROM', '');
  },

  // GRAVATAR
  get AUTH_GRAVATAR_ENABLED() {
    return castBooleanEnv('AUTH_GRAVATAR_ENABLED', true);
  },
  get AUTH_GRAVATAR_DEFAULT() {
    return castStringEnv('AUTH_GRAVATAR_DEFAULT', 'blank');
  },
  get AUTH_GRAVATAR_RATING() {
    return castStringEnv('AUTH_GRAVATAR_RATING', 'g');
  },

  // CLIENT
  get AUTH_CLIENT_URL() {
    return castStringEnv('AUTH_CLIENT_URL', '');
  },

  // SIGN UP
  get AUTH_ANONYMOUS_USERS_ENABLED() {
    return castBooleanEnv('AUTH_ANONYMOUS_USERS_ENABLED', false);
  },
  get AUTH_DISABLE_NEW_USERS() {
    return castBooleanEnv('AUTH_DISABLE_NEW_USERS', false);
  },
  get AUTH_ACCESS_CONTROL_ALLOWED_EMAILS() {
    return castStringArrayEnv('AUTH_ACCESS_CONTROL_ALLOWED_EMAILS', []);
  },
  get AUTH_ACCESS_CONTROL_ALLOWED_EMAIL_DOMAINS() {
    return castStringArrayEnv('AUTH_ACCESS_CONTROL_ALLOWED_EMAIL_DOMAINS', []);
  },
  get AUTH_ACCESS_CONTROL_BLOCKED_EMAILS() {
    return castStringArrayEnv('AUTH_ACCESS_CONTROL_BLOCKED_EMAILS', []);
  },
  get AUTH_ACCESS_CONTROL_BLOCKED_EMAIL_DOMAINS() {
    return castStringArrayEnv('AUTH_ACCESS_CONTROL_BLOCKED_EMAIL_DOMAINS', []);
  },
  get AUTH_PASSWORD_MIN_LENGTH() {
    return castIntEnv('AUTH_PASSWORD_MIN_LENGTH', 3);
  },
  get AUTH_PASSWORD_HIBP_ENABLED() {
    return castBooleanEnv('AUTH_PASSWORD_HIBP_ENABLED', false);
  },
  get AUTH_USER_DEFAULT_ROLE() {
    return castStringEnv('AUTH_USER_DEFAULT_ROLE', 'user');
  },
  get AUTH_USER_DEFAULT_ALLOWED_ROLES() {
    return castStringArrayEnv('AUTH_USER_DEFAULT_ALLOWED_ROLES', [
      'me',
      this.AUTH_USER_DEFAULT_ROLE,
    ]);
  },
  get AUTH_ALLOWED_USER_ROLES() {
    return castStringArrayEnv(
      'ALLOWED_USER_ROLES',
      this.AUTH_USER_DEFAULT_ALLOWED_ROLES
    );
  },
  get AUTH_LOCALE_DEFAULT() {
    return castStringEnv('AUTH_LOCALE_DEFAULT', 'en');
  },
  get AUTH_LOCALE_ALLOWED_LOCALES() {
    return castStringArrayEnv('AUTH_LOCALE_ALLOWED_LOCALES') || ['en'];
  },

  // SIGN IN
  get AUTH_EMAIL_PASSWORDLESS_ENABLED() {
    return castBooleanEnv('AUTH_EMAIL_PASSWORDLESS_ENABLED', false);
  },
  get AUTH_SMS_PASSWORDLESS_ENABLED() {
    return castBooleanEnv('AUTH_SMS_PASSWORDLESS_ENABLED', false);
  },
  get AUTH_EMAIL_SIGNIN_EMAIL_VERIFIED_REQUIRED() {
    return castBooleanEnv('AUTH_EMAIL_SIGNIN_EMAIL_VERIFIED_REQUIRED', true);
  },
  // get AUTH_SIGNIN_PHONE_NUMBER_VERIFIED_REQUIRED() {
  //   return castBooleanEnv('AUTH_SIGNIN_PHONE_NUMBER_VERIFIED_REQUIRED', true);
  // },
  get AUTH_ACCESS_CONTROL_ALLOWED_REDIRECT_URLS() {
    return castStringArrayEnv('AUTH_ACCESS_CONTROL_ALLOWED_REDIRECT_URLS');
  },
  get AUTH_MFA_ENABLED() {
    return castBooleanEnv('AUTH_MFA_ENABLED', false);
  },
  get AUTH_MFA_TOTP_ISSUER() {
    return castStringEnv('AUTH_MFA_TOTP_ISSUER', 'hasura-auth');
  },

  // TOKENS
  get AUTH_ACCESS_TOKEN_EXPIRES_IN() {
    return castIntEnv('AUTH_ACCESS_TOKEN_EXPIRES_IN', 900);
  },
  get AUTH_REFRESH_TOKEN_EXPIRES_IN() {
    return castIntEnv('AUTH_REFRESH_TOKEN_EXPIRES_IN', 43200);
  },

  // EMAIL TEMPLATES
  get AUTH_EMAIL_TEMPLATE_FETCH_URL() {
    return castStringEnv('AUTH_EMAIL_TEMPLATE_FETCH_URL', '');
  },

  // LOGS
  get AUTH_LOG_LEVEL() {
    return castStringEnv('AUTH_LOG_LEVEL', 'info');
  },

  get AUTH_JWT_CUSTOM_CLAIMS() {
    return castObjectEnv<Record<string, string>>('AUTH_JWT_CUSTOM_CLAIMS');
  },

  get AUTH_USER_SESSION_VARIABLE_FIELDS(): Record<string, string> {
    return this.AUTH_USER_SESSION_VARIABLE_FIELDS;
  },
};
