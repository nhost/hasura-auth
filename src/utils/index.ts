export * from './redirect';
export * from './password';
export * from './user';
export * from './qr-code';
export * from './avatar';
export * from './otp';
export * from './gql-sdk';
export * from './ticket';
export * from './tokens';
export * from './jwt';
export * from './env';
export * from './hasura';
export * from './async-wrapper';
