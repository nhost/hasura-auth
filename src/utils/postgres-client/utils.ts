import { User } from '@/types';
import { PoolClient } from 'pg';
import { SqlUser } from './types';

export const cameliseUser = (user?: SqlUser | null): User | null => {
  if (!user) {
    return null;
  }
  const {
    avatar_url,
    created_at,
    default_role,
    display_name,
    email_verified,
    id,
    is_anonymous,
    locale,
    phone_number_verified,
    roles,
    active_mfa_type,
    email,
    metadata,
    phone_number,
    new_email,
    totp_secret,
    disabled,
    ticket,
    password_hash,
    otp_hash,
    webauthn_current_challenge,
    ticket_expires_at,
    otp_method_last_used,
    otp_hash_expires_at,
    last_seen,
  } = user;
  return {
    avatarUrl: avatar_url,
    createdAt: created_at,
    disabled,
    defaultRole: default_role,
    displayName: display_name,
    email,
    emailVerified: email_verified,
    id,
    isAnonymous: is_anonymous,
    locale,
    metadata,
    phoneNumber: phone_number,
    phoneNumberVerified: phone_number_verified,
    roles,
    activeMfaType: active_mfa_type,
    newEmail: new_email,
    totpSecret: totp_secret,
    ticket,
    passwordHash: password_hash,
    otpHash: otp_hash,
    otpMethodLastUsed: otp_method_last_used,
    webauthnCurrentChallenge: webauthn_current_challenge,
    ticketExpiresAt: ticket_expires_at,
    otpHashExpiresAt: otp_hash_expires_at,
    lastSeen: last_seen,
  };
};

export const snakeiseUser = (
  user: Partial<User> | null
): Partial<SqlUser> | null => {
  if (!user) {
    return null;
  }
  const {
    avatarUrl,
    createdAt,
    defaultRole,
    displayName,
    emailVerified,
    id,
    isAnonymous,
    locale,
    phoneNumberVerified,
    roles,
    activeMfaType,
    email,
    metadata,
    phoneNumber,
    newEmail,
    totpSecret,
    disabled,
    ticket,
    passwordHash,
    otpHash,
    webauthnCurrentChallenge,
    ticketExpiresAt,
    otpMethodLastUsed,
    otpHashExpiresAt,
    lastSeen,
  } = user;
  const result: Partial<SqlUser> = {
    avatar_url: avatarUrl,
    created_at: createdAt,
    disabled,
    default_role: defaultRole,
    display_name: displayName,
    email,
    email_verified: emailVerified,
    id,
    is_anonymous: isAnonymous,
    locale,
    metadata,
    phone_number: phoneNumber,
    phone_number_verified: phoneNumberVerified,
    roles,
    active_mfa_type: activeMfaType,
    new_email: newEmail,
    totp_secret: totpSecret,
    ticket,
    password_hash: passwordHash,
    otp_hash: otpHash,
    otp_method_last_used: otpMethodLastUsed,
    webauthn_current_challenge: webauthnCurrentChallenge,
    ticket_expires_at: ticketExpiresAt,
    otp_hash_expires_at: otpHashExpiresAt,
    last_seen: lastSeen,
  };
  Object.keys(result).forEach((k) => {
    const key = k as keyof typeof result;
    if (result[key] === undefined) {
      delete result[key];
    }
  });
  return result;
};

export const createUserQueryWhere = (where: string) =>
  `SELECT u.*, r.roles FROM "auth"."users" u 
    LEFT JOIN lateral (
            SELECT user_id, coalesce(json_agg(role), '[]') AS roles
            FROM "auth"."user_roles" r
            WHERE r.user_id = u.id
            GROUP BY user_id
        ) r ON r.user_id = id WHERE ${where};`;

export const createUserQueryByColumn = (column: string) =>
  createUserQueryWhere(`u.${column} = $1`);

export const getUserById = async (client: PoolClient, userId: string) => {
  const { rows } = await client.query<SqlUser>(createUserQueryByColumn('id'), [
    userId,
  ]);
  return rows[0];
};

export const insertUserRoles = async (
  client: PoolClient,
  userId: string,
  roles: string[]
) => {
  await client.query(
    `INSERT INTO "auth"."user_roles" (user_id, role) VALUES ${roles
      .map((_, i) => `($1, $${i + 2})`)
      .join(', ')} ON CONFLICT DO NOTHING;`,
    [userId, ...roles]
  );
};
