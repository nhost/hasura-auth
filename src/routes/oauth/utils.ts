import { NextFunction, Request, Response } from 'express';
import { GrantConfig, GrantResponse } from 'grant';

import { castBooleanEnv, castStringArrayEnv } from '@config';
import {
  locale as localeValidator,
  email as emailValidator,
} from '@/validation';
import { InsertUserMutationVariables } from '@/utils/__generated__/graphql-request';
import { ENV, getGravatarUrl } from '@/utils';
import { UserRegistrationOptions } from '@/types';

import { OAUTH_ROUTE, PROVIDERS_CONFIG } from './config';

/**
 * Fields that can be possibly returned by the OAuth provider and stored in the database
 */
export type NormalisedProfile = Partial<{
  id: string;
  displayName: string;
  avatarUrl: string;
  email: string;
  locale: string;
  emailVerified: boolean;
}>;

/**
 * Transform the profile normalised from the provider to the format we store in the database.
 * - Options sent by the client (locale, displayName) take precedence over the profile returned by the provider.
 * - Fall back to the email as the display name
 * - the locale should be in the list of allowed locales, and if not, fall back to the default locale
 */
export const transformOauthProfile = async (
  normalised: NormalisedProfile,
  options?: Partial<UserRegistrationOptions>
): Promise<InsertUserMutationVariables['user']> => {
  // * Check if the email is valid. If not, throw an error
  const email = await emailValidator.validateAsync(normalised.email);

  // * Get the avatar URL from the normalised Oauth profile, and fallback to Gravatar if enabled
  const avatarUrl = normalised.avatarUrl || getGravatarUrl(email) || '';

  // * check if the locale is allowed, and if not, use the default one
  let locale = ENV.AUTH_LOCALE_DEFAULT;
  const customLocale = localeValidator.validate(
    options?.locale || normalised.locale
  );
  if (!customLocale.error) {
    locale = customLocale.value;
  }

  /**
   * In order of priority:
   * 1. The value sent as an option by the user
   * 2. The value sent by the Oauth provider
   * 3. The email
   */
  const displayName = options?.displayName || normalised.displayName || email;

  const emailVerified = !!normalised.emailVerified;

  let allowedRoles: string[] = ENV.AUTH_USER_DEFAULT_ALLOWED_ROLES;

  if (options?.allowedRoles) {
    if (Array.isArray(options.allowedRoles)) {
      allowedRoles = options.allowedRoles;
    } else if (typeof options.allowedRoles === 'string') {
      //if for some reason it comes as a string, split it
      allowedRoles = (options.allowedRoles as string).split(',');
    }
  }

  return {
    passwordHash: null,
    metadata: options?.metadata || {},
    email,
    emailVerified,
    defaultRole: options?.defaultRole || ENV.AUTH_USER_DEFAULT_ROLE,
    roles: {
      data: allowedRoles.map((role) => ({
        role,
      })),
    },
    locale,
    displayName,
    avatarUrl,
  };
};

export const normaliseProfile = (provider: string, data: GrantResponse) =>
  PROVIDERS_CONFIG[provider].profile(data);

export const preRequestProviderMiddleware = (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  const provider = req.params.provider;
  const middleware = PROVIDERS_CONFIG[provider]?.middleware;
  if (middleware) {
    return middleware(req, res, next);
  }
  next();
};

/**
 * Create Grant standard configuration
 * - Sets the Grant defauls
 * - Enables the providers defined in the env variables
 * - Adds the custom scope defined in the env variables
 */
export const createGrantConfig = (): GrantConfig =>
  Object.keys(PROVIDERS_CONFIG).reduce<GrantConfig>(
    (aggr, provider) => {
      // * Convert the provider name used in Grant into its name in the env variables
      let providerEnvName = provider.toUpperCase();
      if (providerEnvName === 'WINDOWSLIVE') {
        providerEnvName = 'WINDOWS_LIVE';
      }
      if (castBooleanEnv(`AUTH_PROVIDER_${providerEnvName}_ENABLED`)) {
        const grant = { ...PROVIDERS_CONFIG[provider].grant };
        const customScope = castStringArrayEnv(
          `AUTH_PROVIDER_${providerEnvName}_SCOPE`
        );
        if (customScope.length) {
          // * Adds the scope defined in the env variables to the scope sent as a parameter
          const initialScope = Array.isArray(grant.scope)
            ? grant.scope
            : typeof grant.scope === 'string'
            ? [grant.scope]
            : [];
          // * Merge both sources, and remove duplicates
          grant.scope = [...new Set([...initialScope, ...customScope])];
        }
        aggr[provider] = grant;
      }
      return aggr;
    },
    {
      defaults: {
        prefix: `${ENV.AUTH_API_PREFIX}${OAUTH_ROUTE}`,
        transport: 'session',
        scope: ['email', 'profile'],
        response: ['tokens', 'email', 'profile', 'jwt'],
      },
    }
  );
