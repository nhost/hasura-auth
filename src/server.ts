import { ENV } from '@/utils/env';
import { logger, LOG_LEVEL } from './logger';

export const start = async () => {
  logger.info(`Log level: ${LOG_LEVEL}`);

  /**
   * Async imports allow to dynamically import the required modules,
   * and therefore only import modules that are required, when they are required.
   * It decreases the loading time on startup.
   * This distinction can make a difference if using hasura-auth to only apply migrations/metadata,
   * or to skip migrations/metadata on startup.
   */
  // if (ENV.AUTH_SKIP_INIT) {
  //   logger.info(`Skipping migrations and metadata`);
  // } else {
  const { waitForHasura } = await import('@/utils');
  const { applyMigrations } = await import('./migrations');
  const { applyMetadata } = await import('./metadata');

  // wait for hasura to be ready
  await waitForHasura();
  // apply migrations and metadata
  await applyMigrations();
  await applyMetadata();
  // }

  // if (ENV.AUTH_SKIP_SERVE) {
  //   logger.info(
  //     `AUTH_SKIP_SERVE has been set to true, therefore Hasura-Auth won't start`
  //   );
  // } else {
  await import('./env-vars-check');
  const { app } = await import('./app');

  app.listen(ENV.AUTH_PORT, () => {
    logger.info(`Running on port ${ENV.AUTH_PORT}`);
  });
  // }
};
