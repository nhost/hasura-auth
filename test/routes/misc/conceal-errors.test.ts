import { Client } from 'pg';
import * as faker from 'faker';
import { StatusCodes } from 'http-status-codes';

import { ENV } from '../../../src/utils/env';
import { request } from '../../server';
import { deleteAllMailHogEmails } from '../../utils';

describe('conceal error messages', () => {
  let client: Client;

  beforeAll(async () => {
    // set env vars
    await request.post('/change-env').send({
      AUTH_DISABLE_NEW_USERS: false,
      AUTH_EMAIL_SIGNIN_EMAIL_VERIFIED_REQUIRED: false,
      AUTH_ANONYMOUS_USERS_ENABLED: false,
    });
    client = new Client({
      connectionString: ENV.HASURA_GRAPHQL_DATABASE_URL,
    });
    await client.connect();
    await deleteAllMailHogEmails();
  });

  afterAll(async () => {
    await request.post('/change-env').send({
      AUTH_CONCEAL_ERRORS: false,
    });
    await client.end();
  });

  beforeEach(async () => {
    await client.query(`DELETE FROM auth.users;`);
  });

  it('should conceal errors', async () => {
    const email = faker.internet.email();
    const password = faker.internet.password();
    const user = {
      email,
      password,
    };

    await request
      .post('/signup/email-password')
      .send(user)
      .expect(StatusCodes.OK);

    // * Conceal error messages
    await request.post('/change-env').send({
      AUTH_CONCEAL_ERRORS: true,
    });

    const attemptWithProtection = await request
      .post('/signin/email-password')
      .send({ email, password: '123456' });

    expect(attemptWithProtection.body).toMatchSnapshot();

    // * Send details error messages
    await request.post('/change-env').send({
      AUTH_CONCEAL_ERRORS: false,
    });

    const attemptWithoutProtection = await request
      .post('/signin/email-password')
      .send({ email, password: '123456' });

    expect(attemptWithoutProtection.body).toMatchSnapshot();
  });
});
