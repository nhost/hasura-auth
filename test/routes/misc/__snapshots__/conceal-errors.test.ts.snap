// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conceal error messages should conceal errors 1`] = `
Object {
  "error": "invalid-email-password",
  "message": "The request payload is incorrect",
  "status": 400,
}
`;

exports[`conceal error messages should conceal errors 2`] = `
Object {
  "error": "invalid-email-password",
  "message": "Incorrect email or password",
  "status": 401,
}
`;
